utils/github_reporting.py

```diff
@@ -0,0 +1,531 @@
+        output_string = "{}\n" \
+                        "Complete test suite and dependencies: {}\n" \
+                        "## Failed tests\n".format(COMMENT_TITLE,
+                                                   self.get_dependency_url())
+        number_of_failures = 0
+
+        for k in results:
+            if results[k]:
+                output_string += '* ' + str(k) + "\n"
+                for fail in results[k]:
+                    output_string += '  * ' + str(fail) + '\n'
+                    number_of_failures += 1
+
+        if not number_of_failures:
+            output_string += "No failures!\n"
+
+        return output_string
+
+    def __str__(self):
+        return self.job_name
+
+
+class GitHubRepo:
+    def __init__(self, repo_name, owner="QubesOS"):
+        self.repo = repo_name
+        self.url = "{}/{}/{}/". format(GITHUB_API_PREFIX, owner, self.repo)
+
+    def get_issues_by_name(self, name):
+        url = self.url + 'issues'
+        while url:
+            data = requests.get(url, headers=github_auth)
+            json_data = data.json()
+
+            for issue in json_data:
+                if issue['title'] == name:
+                    return issue['html_url']
+
+            if 'next' in data.links.keys():
+                url = data.links['next']['url']
+            else:
+                url = None
+
+
+class GitHubIssue:
+    def __init__(self, url):
+        self.url, self.issue_no = self.parse_url(url)
+        self.init_github_auth()
+
+    @staticmethod
+    def init_github_auth():
+        if not github_auth:
+            assert 'GITHUB_API_KEY' in os.environ
+            github_auth['Authorization'] = \
+                'token {}'.format(os.environ['GITHUB_API_KEY'])
+
+    def existing_comment(self):
+        comments_url = self.url + '{}/comments'.format(self.issue_no)
+        while comments_url:
+            comments_data = requests.get(comments_url, headers=github_auth)
+            comments_json = comments_data.json()
+
+            for comment in comments_json:
+                comment_title = comment['body'][:len(COMMENT_TITLE)]
+                if comment_title == COMMENT_TITLE:
+                    return comment['id']
+
+            if 'next' in comments_data.links.keys():
+                comments_url = comments_data.links['next']['url']
+            else:
+                comments_url = None
+
+        return None
+
+    @staticmethod
+    def parse_url(url):
+        pattern = r'{}/(\w+)/([\w\-]+)/(issues|pull)/(\d+)#?.*'.format(
+            GITHUB_BASE_PREFIX)
+        match = re.compile(pattern).match(url)
+
+        owner = match.group(1)
+        repo = match.group(2)
+        no = match.group(4)
+
+        parsed_url = "{}/{}/{}/issues/".format(
+                GITHUB_API_PREFIX,
+                owner,
+                repo,
+                no)
+
+        return parsed_url, no
+
+    def post_comment(self, message_text):
+        if self.existing_comment():
```

Comment:
`self.existing_comment()` potentially makes multiple API calls, please cache the result, as you use it a line below. Either here, or in the method itself.